datafusion/sql/src/query.rs

```diff
@@ -283,3 +276,44 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         }
     }
 }
+
+/// Retrieves the constant result of an expression, evaluating it if possible.
+///
+/// This function takes an expression and an argument name as input and returns
+/// a `Result<i64>` indicating either the constant result of the expression or an
+/// error if the expression cannot be evaluated.
+///
+/// # Arguments
+///
+/// * `expr` - An `Expr` representing the expression to evaluate.
+/// * `arg_name` - The name of the argument for error messages.
+///
+/// # Returns
+///
+/// * `Result<i64>` - An `Ok` variant containing the constant result if evaluation is successful,
+///   or an `Err` variant containing an error message if evaluation fails.
+fn get_constant_result(expr: &Expr, arg_name: &str) -> Result<i64> {
+    match expr {
```

Comment:
We could use `ConstEvaluator` and see if it is evaluated to a constant value

Reply:
To use this helper. I have imported `datafusion-optimizer` to `datafusion-sql` crate. This dependency caused an error in `test_deps` test. Hence I  am retracting these changes for now.

Reply:
Could we change the `Limit` logical plan to support arbitrary expressions? 
```rust
pub struct Limit {
    pub skip: Expr,
    pub fetch: Option<Expr>,
    pub input: Arc<LogicalPlan>,
}
```
The `SimplifyExpressions` rule can automatically optimize them into constants. Some optimization rules such as `PushDownLimit` only run when the limit expression is a constant. We may need to add a cast for the limit expression, only checking if it is a constant of type u64.

When creating the `LimitExec` physical plan, convert the limit expression into `PhysicalExpr` and evaluate it.

Reply:
> Could we change the Limit logical plan to support arbitrary expressions?

I agree this would be the "correct" way to support arbitrarily simplifiable expressions in the limit clause

However, I suspect it might be a major change

So I guess it comes down to "how important is the `+` / `-` case -- if it is really important then we can proceed with this PR, but if the real use case is arbitrary expressions, it probably makes sense to do the more substantial change as suggested by @jonahgao

Reply:
I have experimented with the keeping `Expr` in the limit state. However, it is not a trivial change. Also, it is not clear when to do error checking, when to ignore errors while converting  `skip` and `fetch` to their corresponding `u64` versions. I suggest, we first merge this PR. Then, we can have a support for arbitrary expressions in another PR (Also I think, we can postpone this feature until a use case hits.).

Reply:
I filed https://github.com/apache/arrow-datafusion/issues/9821 to track